### ISSUE TITLE
sudoers: the runas denylist scan is escape-blind for `(`, `)`, `>` and `"`, drawing four false `sudo-F02`s

### ISSUE THREAD

**Issue by ErstBlack** [rulesteward/rulesteward#700]
## Summary

`first_invalid_char` (`crates/rulesteward-sudoers/src/lints/tokens/runas.rs:39`) scans a
runas token for any char in `is_denylist_char` (`lints/tokens/shared.rs:11`, the set
`! ( ) > "`) and reports `sudo-F02` on a hit. The scan compares the char alone, so a
BACKSLASH-ESCAPED denylist char is reported as invalid even though `visudo` accepts it as a
literal character in the principal name.

#699 fixes the `!` member of that set, because it is entangled with the negation-sigil
boundary faces and its parity control points the opposite way. **This issue is the other
four**, which are a plain uniform case and were left out of #699 deliberately rather than
overlooked.

The diagnostic text makes the claim explicit and false: *"contains an invalid character `>`
that visudo rejects"*, on a file visudo accepts.

## Ground truth

Re-derived 2026-08-19 on `rs-oracle9` (sudo 1.9.17p2), stdin only, `--network=none`.
Every accepted row consumes the escape and keeps the bare char in the username.

| input | visudo | cvtsudoers runasusers | RuleSteward today |
|---|---|---|---|
| `alice ALL = (a\(b) /bin/ls` | **rc 0** | `[{"username":"a(b"}]` | **false `sudo-F02`** on `` ` ` `` `(` |
| `alice ALL = (a\)b) /bin/ls` | **rc 0** | `[{"username":"a)b"}]` | **false `sudo-F02`** on `)` |
| `alice ALL = (a\>b) /bin/ls` | **rc 0** | `[{"username":"a>b"}]` | **false `sudo-F02`** on `>` |
| `alice ALL = (a\"b) /bin/ls` | **rc 0** | `[{"username":"a\"b"}]` | **false `sudo-F02`** on `"` |
| `alice ALL = (a>b) /bin/ls` | rc 1 | - | `sudo-F02`, CORRECT (the control) |

## Severity

Lower than #699. This is a FALSE FATAL, not a fail-open: it reports a problem that is not
there rather than staying silent on one that is. It still matters, because #668's own
framing applies in reverse - a false FATAL on a legal file trains an operator to ignore
`sudo-F02`, and `sudo-F02` is the code that catches the genuine `(ro!ot)` and `(a>b)`
rejects.

## Fix sketch

Make the scan ask the SEPARATOR rule, exactly as #699 does for `!`:

```rust
bare.char_indices()
    .find(|&(i, c)| {
        (is_denylist_char(c) || c.is_whitespace()) && !separator_escaped(bare, i)
    })
    .map(|(_, c)| c)
```

Two things a fix must not do:

1. **Do not drop chars from `is_denylist_char`.** All five are genuinely invalid UNESCAPED,
   and the set is shared with `check_group_subject`. `(a>b)` must keep firing.
2. **Do not assume the whitespace half behaves the same.** The `|| c.is_whitespace()`
   disjunct in the same expression has its own grounding (#652 face B) and an escaped space
   has NOT been probed here. Derive `(a\ b)` on the oracle before deciding whether the
   escape-awareness covers it too, and record the answer either way.

Also re-check `clean_double_quoted_interior`'s early return above the scan: with the `"`
member escape-aware, an escaped quote inside an otherwise-clean region may reach a different
branch than it does today.

## Sweep

`is_denylist_char` has two callers: `runas.rs::first_invalid_char` (this issue) and the
subject-position sub-case (b) check in `check_user_spec`. The subject-position caller was
NOT probed for the escaped axis in 2026-08-19's sigil sweep, so it needs its own
`visudo`/`cvtsudoers` rows before it can be marked clean or fixed. Do not assume it mirrors
the runas answer: #699's parity controls showed the same byte sequence getting opposite
verdicts in runas vs user-list position.

**Comment by ErstBlack** [comment]
## The whitespace half is now probed, and this issue's fix sketch covers it

This issue's "Fix sketch" asked a future session to derive `(a\ b)` on the oracle before
deciding whether the escape-awareness extends to the `|| c.is_whitespace()` disjunct, and
explicitly warned against assuming it behaves like `!`. Derived 2026-08-19 on `rs-oracle9`
(sudo 1.9.17p2, stdin, `--network=none`), so the answer is on record rather than re-asked:

| input | `visudo` | `cvtsudoers` runasusers | RuleSteward |
|---|---|---|---|
| `alice ALL = (a\ b) /bin/ls` | **rc 0** | `[{"username":"a b"}]` | **false `sudo-F02`** |
| `alice ALL = (a\\ b) /bin/ls` | rc 1 | - | control: even run, literal backslash, real separator |

So the whitespace disjunct is escape-blind in exactly the same shape as the four denylist
chars, and its parity control points the same way as `!`'s. The sketch's
`!separator_escaped(bare, i)` applied to the whole predicate does cover it; no separate
treatment is needed, and the caveat in this issue's body can be resolved rather than
carried.

Three further rows from the same sweep, all `visudo` **rc 0** and all drawing a false
`sudo-F02` today, which belong to this issue rather than to #699 (they are unescaped, so
#699's fix does not reach them):

| input | `visudo` | `cvtsudoers` runasusers |
|---|---|---|
| `alice ALL = (! a) /bin/ls` | rc 0 | `[{"username":"a","negated":true}]` |
| `alice ALL = (a!!) /bin/ls` | rc 0 | `[{"username":"a"}]` |
| `alice ALL = (\\!!) /bin/ls` | rc 0 | `[{"username":"\\"}]` |
| `alice ALL = (a!) /bin/ls` | rc 1 | control, so `!!` really does cancel |

29 such rows appeared in an exhaustive length-<=4 runas sweep. Note `(! a)`: a sigil
SEPARATED from its principal by a blank is still a negation, which the current
`trim_start_matches('!')` cannot express.

## Related, filed since

**#703** covers the opposite direction on the same function: `(!)` and `("")` return "clean"
on tokens `visudo` REJECTS. Worth fixing together with this issue, since both land in
`first_invalid_char` and a repair of one can mask the other - `(!)` trims to an empty string,
and any escape-awareness added here runs after that trim.

**#699** is merged into this branch and fixed only the `!` member, deliberately, with the
narrow scope named in the code comment so the gap is auditable rather than unexamined.
